// fundamental-ngx/esm5/lib/menu/menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * The component that represents a menu.
 */
var MenuComponent = /** @class */ (function () {
    function MenuComponent() {
        /**
         * @hidden
         */
        this.fdMenuClass = true;
    }
    MenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-menu',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".fd-menu__item:focus{outline:var(--fd-color-action-focus) dotted 1px}"]
                }] }
    ];
    MenuComponent.propDecorators = {
        fdMenuClass: [{ type: HostBinding, args: ['class.fd-menu',] }]
    };
    return MenuComponent;
}());
export { MenuComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    MenuComponent.prototype.fdMenuClass;
}
