// fundamental-ngx/esm5/lib/token/token.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Output, ViewChild, ViewEncapsulation } from '@angular/core';
/**
 * A token is used to represent contextualizing information.
 * They can be useful to show applied filters, selected values for form fields or object metadata.
 */
var TokenComponent = /** @class */ (function () {
    /** @hidden */
    function TokenComponent(elRef) {
        this.elRef = elRef;
        /**
         * Emitted when the *x* icon is clicked. Specifically, any pseudo-element.
         */
        this.onCloseClick = new EventEmitter();
    }
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    TokenComponent.prototype.clickHandler = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.contentContainer) {
            if (this.elRef.nativeElement.contains(event.target) && !this.contentContainer.nativeElement.contains(event.target)) {
                this.onCloseClick.emit();
            }
        }
    };
    TokenComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-token',
                    template: "<span class=\"fd-token-content\" #contentContainer>\n    <ng-content></ng-content>\n</span>\n\n",
                    host: {
                        class: 'fd-token',
                        'role': 'button'
                    },
                    encapsulation: ViewEncapsulation.None,
                    styles: [".fd-token-content{display:inline-block}"]
                }] }
    ];
    /** @nocollapse */
    TokenComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    TokenComponent.propDecorators = {
        contentContainer: [{ type: ViewChild, args: ['contentContainer',] }],
        onCloseClick: [{ type: Output }],
        clickHandler: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return TokenComponent;
}());
export { TokenComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    TokenComponent.prototype.contentContainer;
    /**
     * Emitted when the *x* icon is clicked. Specifically, any pseudo-element.
     * @type {?}
     */
    TokenComponent.prototype.onCloseClick;
    /**
     * @type {?}
     * @private
     */
    TokenComponent.prototype.elRef;
}
