// fundamental-ngx/esm5/lib/modal/modal-service/modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Injectable, ApplicationRef, Injector } from '@angular/core';
import { ModalComponent } from '../modal.component';
import { ModalBackdrop } from '../modal-utils/modal-backdrop';
import { ModalContainer } from '../modal-utils/modal-container';
import { ModalConfig } from '../modal-utils/modal-config';
import { ModalRef } from '../modal-utils/modal-ref';
import { ModalInjector } from '../modal-utils/modal-injector';
/**
 * Service used to dynamically generate a modal.
 */
var ModalService = /** @class */ (function () {
    /** @hidden */
    function ModalService(componentFactoryResolver, appRef, injector) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
        this.modals = [];
    }
    /**
     * Status of the modal service.
     * Returns true if there are open modals, false otherwise.
     */
    /**
     * Status of the modal service.
     * Returns true if there are open modals, false otherwise.
     * @return {?}
     */
    ModalService.prototype.hasOpenModals = /**
     * Status of the modal service.
     * Returns true if there are open modals, false otherwise.
     * @return {?}
     */
    function () {
        return this.modals && this.modals.length > 0;
    };
    /**
     * Dismisses all currently open modals.
     */
    /**
     * Dismisses all currently open modals.
     * @return {?}
     */
    ModalService.prototype.dismissAll = /**
     * Dismisses all currently open modals.
     * @return {?}
     */
    function () {
        var _this = this;
        this.modals.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.destroyModalComponent(item.modalRef);
        }));
    };
    /**
     * Opens a modal component with a content of type TemplateRef or a component type.
     * @param contentType Content of the modal component.
     * @param modalConfig Configuration of the modal component.
     */
    /**
     * Opens a modal component with a content of type TemplateRef or a component type.
     * @param {?} contentType Content of the modal component.
     * @param {?=} modalConfig Configuration of the modal component.
     * @return {?}
     */
    ModalService.prototype.open = /**
     * Opens a modal component with a content of type TemplateRef or a component type.
     * @param {?} contentType Content of the modal component.
     * @param {?=} modalConfig Configuration of the modal component.
     * @return {?}
     */
    function (contentType, modalConfig) {
        var _this = this;
        if (modalConfig === void 0) { modalConfig = new ModalConfig(); }
        // Get default values from model
        modalConfig = Object.assign(new ModalConfig(), modalConfig);
        // Setup injectable data
        /** @type {?} */
        var configMap = new WeakMap();
        /** @type {?} */
        var modalRef = new ModalRef();
        modalRef.data = (modalConfig ? modalConfig.data : undefined);
        configMap.set(ModalRef, modalRef);
        // Prepare container
        /** @type {?} */
        var containerFactory = this.componentFactoryResolver.resolveComponentFactory(ModalContainer);
        /** @type {?} */
        var containerRef = containerFactory.create(this.injector);
        this.appRef.attachView(containerRef.hostView);
        // Prepare modal
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(ModalComponent);
        /** @type {?} */
        var componentRef = componentFactory.create(new ModalInjector(this.injector, configMap));
        this.appRef.attachView(componentRef.hostView);
        // Prepare backdrop
        /** @type {?} */
        var backdropRef;
        if (modalConfig.hasBackdrop) {
            /** @type {?} */
            var backdropFactory = this.componentFactoryResolver.resolveComponentFactory(ModalBackdrop);
            backdropRef = backdropFactory.create(new ModalInjector(this.injector, configMap));
            this.appRef.attachView(backdropRef.hostView);
        }
        // Subscribe to close of modalRef
        /** @type {?} */
        var refSub = modalRef.afterClosed.subscribe((/**
         * @return {?}
         */
        function () {
            _this.destroyModalComponent(componentRef);
            refSub.unsubscribe();
        }), (/**
         * @return {?}
         */
        function () {
            _this.destroyModalComponent(componentRef);
            refSub.unsubscribe();
        }));
        // Assign component attributes
        /** @type {?} */
        var configObj = Object.assign({}, modalConfig);
        Object.keys(configObj).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            if (key !== 'data') {
                componentRef.instance[key] = configObj[key];
                if (modalConfig.hasBackdrop) {
                    backdropRef.instance[key] = configObj[key];
                }
            }
        }));
        componentRef.instance.childComponentType = contentType;
        // Sizing
        this.setModalSize(componentRef, configObj);
        // Positioning
        this.setModalPosition(componentRef, configObj.position);
        // Render container
        /** @type {?} */
        var containerEl = (/** @type {?} */ (((/** @type {?} */ (containerRef.hostView))).rootNodes[0]));
        if (modalConfig.container !== 'body') {
            modalConfig.container.appendChild(containerEl);
        }
        else {
            document.body.appendChild(containerEl);
        }
        // Render backdrop
        if (modalConfig.hasBackdrop) {
            /** @type {?} */
            var domBackdrop = (/** @type {?} */ (((/** @type {?} */ (backdropRef.hostView))).rootNodes[0]));
            containerRef.location.nativeElement.appendChild(domBackdrop);
        }
        // Render modal
        /** @type {?} */
        var domElement = (/** @type {?} */ (((/** @type {?} */ (componentRef.hostView))).rootNodes[0]));
        containerRef.location.nativeElement.appendChild(domElement);
        this.modals.push({ modalRef: componentRef, backdropRef: backdropRef, containerRef: containerRef });
        return modalRef;
    };
    /**
     * @private
     * @param {?} modal
     * @return {?}
     */
    ModalService.prototype.destroyModalComponent = /**
     * @private
     * @param {?} modal
     * @return {?}
     */
    function (modal) {
        /** @type {?} */
        var arrayRef = this.modals.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.modalRef === modal; }));
        /** @type {?} */
        var indexOf = this.modals.indexOf(arrayRef);
        this.appRef.detachView(arrayRef.modalRef.hostView);
        this.appRef.detachView(arrayRef.containerRef.hostView);
        arrayRef.containerRef.destroy();
        arrayRef.modalRef.destroy();
        if (arrayRef.backdropRef) {
            this.appRef.detachView(arrayRef.backdropRef.hostView);
            arrayRef.backdropRef.destroy();
        }
        this.modals[indexOf] = null;
        this.modals = this.modals.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item !== null && item !== undefined; }));
    };
    /**
     * @private
     * @param {?} componentRef
     * @param {?} configObj
     * @return {?}
     */
    ModalService.prototype.setModalSize = /**
     * @private
     * @param {?} componentRef
     * @param {?} configObj
     * @return {?}
     */
    function (componentRef, configObj) {
        componentRef.location.nativeElement.style.minWidth = configObj.minWidth;
        componentRef.location.nativeElement.style.minHeight = configObj.minHeight;
        componentRef.location.nativeElement.style.maxWidth = configObj.maxWidth;
        componentRef.location.nativeElement.style.maxHeight = configObj.maxHeight;
        componentRef.location.nativeElement.style.width = configObj.width;
        componentRef.location.nativeElement.style.height = configObj.height;
    };
    /**
     * @private
     * @param {?} componentRef
     * @param {?} position
     * @return {?}
     */
    ModalService.prototype.setModalPosition = /**
     * @private
     * @param {?} componentRef
     * @param {?} position
     * @return {?}
     */
    function (componentRef, position) {
        if (position) {
            componentRef.location.nativeElement.style.top = position.top;
            componentRef.location.nativeElement.style.bottom = position.bottom;
            componentRef.location.nativeElement.style.right = position.right;
            componentRef.location.nativeElement.style.left = position.left;
        }
    };
    ModalService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ModalService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ApplicationRef },
        { type: Injector }
    ]; };
    return ModalService;
}());
export { ModalService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modals;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.injector;
}
