// fundamental-ngx/esm2015/lib/side-navigation/side-navigation-item/side-navigation-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ViewEncapsulation } from '@angular/core';
import { SideNavigationLinkDirective } from '../side-navigation-link/side-navigation-link.directive';
import { SideNavigationSublistDirective } from '../side-navigation-sublist/side-navigation-sublist.directive';
/**
 * The component that represents a navigation group.
 * ```html
 * <fd-side-nav>
 *    <fd-side-nav-group>
 *        <h1 fd-side-nav-title>Group Name</h1>
 *          <div fd-side-nav-list>
 *             <fd-side-nav-item>
 *                <a fd-side-nav-link [attr.href]="'#'">Link Item</a>
 *             </fd-side-nav-item>
 *          </div>>
 *    </fd-side-nav-group>
 * </fd-side-nav>
 * ```
 */
export class SideNavigationItemComponent {
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this.linkElement && this.subListElement) {
            /** After view content check if there is flag with opened true */
            this.subListElement.subListIsOpenChange(this.linkElement.sublistIsOpen);
            this.subListOpenChanged$ = this.linkElement.onSubListOpenChange.subscribe((/**
             * @param {?} isOpen
             * @return {?}
             */
            isOpen => {
                this.subListElement.subListIsOpenChange(isOpen);
            }));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subListOpenChanged$) {
            this.subListOpenChanged$.unsubscribe();
        }
    }
}
SideNavigationItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-side-nav-item',
                template: "<div class=\"fd-side-nav__item\">\n  <ng-content select=\"[fd-side-nav-link]\"></ng-content>\n  <ng-content select=\"[fd-side-nav-sublist]\"></ng-content>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
SideNavigationItemComponent.propDecorators = {
    linkElement: [{ type: ContentChild, args: [SideNavigationLinkDirective,] }],
    subListElement: [{ type: ContentChild, args: [SideNavigationSublistDirective,] }]
};
if (false) {
    /** @type {?} */
    SideNavigationItemComponent.prototype.linkElement;
    /** @type {?} */
    SideNavigationItemComponent.prototype.subListElement;
    /**
     * \@Hidden
     * @type {?}
     */
    SideNavigationItemComponent.prototype.subListOpenChanged$;
}
