// fundamental-ngx/esm5/lib/side-navigation/side-navigation-link/side-navigation-link.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, ElementRef, Output, EventEmitter, Directive, HostListener, Inject, HostBinding } from '@angular/core';
import { AbstractFdNgxClass } from '../../utils/abstract-fd-ngx-class';
/**
 * The component that represents a navigation link.
 * ```html
 *    <a fd-side-nav-link>
 *        <a [attr.href]="'#'">Link Item</a>
 *    </a>
 * ```
 */
var SideNavigationLinkDirective = /** @class */ (function (_super) {
    tslib_1.__extends(SideNavigationLinkDirective, _super);
    /** @hidden */
    function SideNavigationLinkDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Whether the link has a sublist.
         */
        _this.hasSublist = false;
        /**
         * Whether the sub list is opened or closed
         */
        _this.onSubListOpenChange = new EventEmitter();
        _this.sublistIsOpen = false;
        _this.role = _this.hasSublist ? 'button' : '';
        _this.hasPopup = _this.hasSublist;
        _this.tabindex = _this.hasSublist ? '0' : '';
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    SideNavigationLinkDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-side-nav__link');
        if (this.hasSublist) {
            this._addClassToElement('has-child');
            this._addStyleToElement('cursor', 'pointer');
            this.elementRef.nativeElement.setAttribute('aria-expanded', this.sublistIsOpen);
        }
        if (this.sublistIsOpen && this.hasSublist) {
            this._addClassToElement('is-selected');
        }
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    SideNavigationLinkDirective.prototype.onKeypressHandler = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.hasSublist && (event.code === 'Enter' || event.code === 'Space')) {
            event.preventDefault();
            this.changeSubListIsOpen();
        }
    };
    /**
     * @return {?}
     */
    SideNavigationLinkDirective.prototype.changeSubListIsOpen = /**
     * @return {?}
     */
    function () {
        if (this.hasSublist) {
            this.sublistIsOpen = !this.sublistIsOpen;
            this.onSubListOpenChange.emit(this.sublistIsOpen);
            this.ngOnChanges();
        }
    };
    SideNavigationLinkDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-side-nav-link]',
                },] }
    ];
    /** @nocollapse */
    SideNavigationLinkDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    SideNavigationLinkDirective.propDecorators = {
        hasSublist: [{ type: Input }],
        onSubListOpenChange: [{ type: Output }],
        sublistIsOpen: [{ type: Input }],
        role: [{ type: HostBinding, args: ['attr.role',] }],
        hasPopup: [{ type: HostBinding, args: ['attr.aria-haspopup',] }],
        tabindex: [{ type: HostBinding, args: ['tabindex',] }],
        onKeypressHandler: [{ type: HostListener, args: ['keypress', ['$event.target'],] }],
        changeSubListIsOpen: [{ type: HostListener, args: ['click', ['$event.target'],] }]
    };
    return SideNavigationLinkDirective;
}(AbstractFdNgxClass));
export { SideNavigationLinkDirective };
if (false) {
    /**
     * Whether the link has a sublist.
     * @type {?}
     */
    SideNavigationLinkDirective.prototype.hasSublist;
    /**
     * Whether the sub list is opened or closed
     * @type {?}
     */
    SideNavigationLinkDirective.prototype.onSubListOpenChange;
    /** @type {?} */
    SideNavigationLinkDirective.prototype.sublistIsOpen;
    /** @type {?} */
    SideNavigationLinkDirective.prototype.role;
    /** @type {?} */
    SideNavigationLinkDirective.prototype.hasPopup;
    /** @type {?} */
    SideNavigationLinkDirective.prototype.tabindex;
    /**
     * @type {?}
     * @private
     */
    SideNavigationLinkDirective.prototype.elementRef;
}
