// fundamental-ngx/esm5/lib/side-navigation/side-navigation-sublist/side-navigation-sublist.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
/**
 * The component that represents a navigation group.
 * ```html
 * <fd-side-nav>
 *    <fd-side-nav-group>
 *        <h1 fd-side-nav-title>Group Name</h1>
 *          <div fd-side-nav-list>
 *             <fd-side-nav-item>
 *                <a fd-side-nav-link [attr.href]="'#'">Link Item</a>
 *             </fd-side-nav-item>
 *          </div>>
 *    </fd-side-nav-group>
 * </fd-side-nav>
 * ```
 */
var SideNavigationSublistDirective = /** @class */ (function () {
    /** @hidden */
    function SideNavigationSublistDirective(elementRef) {
        this.elementRef = elementRef;
        this.sublistIsOpen = false;
    }
    /**
     * @param {?} sublistIsOpen
     * @return {?}
     */
    SideNavigationSublistDirective.prototype.subListIsOpenChange = /**
     * @param {?} sublistIsOpen
     * @return {?}
     */
    function (sublistIsOpen) {
        this.sublistIsOpen = sublistIsOpen;
        this.elementRef.nativeElement.setAttribute('aria-hidden', !this.sublistIsOpen);
    };
    SideNavigationSublistDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-side-nav-sublist]',
                    host: {
                        class: 'fd-side-nav__sublist'
                    }
                },] }
    ];
    /** @nocollapse */
    SideNavigationSublistDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return SideNavigationSublistDirective;
}());
export { SideNavigationSublistDirective };
if (false) {
    /** @type {?} */
    SideNavigationSublistDirective.prototype.sublistIsOpen;
    /**
     * @type {?}
     * @private
     */
    SideNavigationSublistDirective.prototype.elementRef;
}
